dusty/systems/virtualbox/__init__.py

```diff
@@ -199,29 +200,55 @@ def get_docker_vm_ip():
-        # get inside the VM and determine the VM's local address on
-        # the virtualized network. This IP is what we're ultimately
-        # trying to return from this function.
-        ip_addr = check_output_demoted(['ssh', '-o', 'StrictHostKeyChecking=no',
-                                        '-o', 'UserKnownHostsFile=/dev/null',
-                                        '-i', _vm_key_path(), '-p', ssh_port,
-                                        'docker@127.0.0.1', 'ip addr show dev eth1'])
-        for line in ip_addr.splitlines():
-            line = line.strip()
-            if line.startswith('inet') and not line.startswith('inet6'):
-                ip = line.split(' ')[1].split('/')[0]
-                return ip
-
-    # If all our crazy hacks up there don't work, we can still get the IP
-    # from Machine directly. The reason we don't do this normally is
-    # because it's really slow, taking around 600ms on my fast machine.
-    logging.warning('Could not get IP the fast way, falling back to Docker Machine')
-    return check_output_demoted(['docker-machine', 'ip', constants.VM_MACHINE_NAME]).rstrip()
+                return host_port
+    raise ValueError('Could not determine localhost port for SSH forwarding')
+
+def _get_host_only_mac_address():
+    # Get the number of the host-only adapter
+    vm_config = _get_vm_config()
+    for line in vm_config:
+        if line.startswith('hostonlyadapter'):
+            adapter_number = int(line[15:16])
+            break
+    else:
+        raise ValueError('No host-only adapter is defined for the Dusty VM')
+
+    for line in vm_config:
+        if line.startswith('macaddress{}'.format(adapter_number)):
+            return line.split('=')[1].strip('"').lower()
+    raise ValueError('Could not find MAC address for adapter number {}'.format(adapter_number))
+
+def _ip_for_mac_from_ip_addr(ip_addr, target_mac):
+    return_next_ip = False
+    for line in ip_addr.splitlines():
+        line = line.strip()
+        if line.startswith('link/ether'):
+            line_mac = line.split(' ')[1].replace(':', '')
+            if line_mac == target_mac:
+                return_next_ip = True
+        elif return_next_ip and line.startswith('inet') and not line.startswith('inet6'):
+            ip = line.split(' ')[1].split('/')[0]
+            return ip
+
+def _get_host_only_ip():
+    """Determine the host-only IP of the Dusty VM through Virtualbox and SSH
+    directly, bypassing Docker Machine. We do this because Docker Machine is
+    much slower, taking about 600ms total. We are basically doing the same
+    flow Docker Machine does in its own code."""
+    mac = _get_host_only_mac_address()
+    ip_addr = check_output_demoted(['ssh', '-o', 'StrictHostKeyChecking=no',
+                                    '-o', 'UserKnownHostsFile=/dev/null',
+                                    '-i', _vm_key_path(), '-p', _get_localhost_ssh_port(),
```

Comment:
No, there's a localhost port forwarded to port 22 in the VM. We can use that to SSH to the VM before we know its actual IP address.